### PR TITLE
Nexus: add example for specifying estimators

### PR DESCRIPTION
**Pull request by aannabe**
## Proposed changes

As discussed previously in persion, a comprehensive example was needed for Nexus that demonstrates how to specify various estimators. Here, I have included Density, SpinDenstiy, EnergyDensity, and 1RDM observables in a single Nexus workflow. The goal is to add all the other estimators too as they mature.

In addition, I added a `scripts_postproc` directory here that includes some post-processing tools. Currently, I included the 1RDM retrieval and plotting tools adapted from discussions in #1052. For now I think it will be helpful to users, at least as a starting point. Let me know if it's not well-suited for here and I will remove it.

For long run, there should be a centralized place for post-processing tools that analyze the estimators in the stat.h5 file. Whether it should be one large tool that is one-fits-all, or multiple specialized smaller tools could also be discussed. Currently, it seems to be not well handled - for instance, I see that `qdens` breaks if 1RDM or energy density is present in the same stat.h5 file. I will create an issue for this if the fix is not straightforward.

## What type(s) of changes does this code introduce?
- New example

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
Local machine

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Review by prckent — COMMENTED** [review]
Please add the pseudopotential files so that this can eventually be automated and success/failure tracked.

**Comment by aannabe** [comment]
Pseudos now added.

**Comment by aannabe** [comment]
Note: I tested that spin-density, 1RDM, and energy-density runs in the batched code with the above Nexus inputs. However, the `density` estimator crashes with the following error: `Fatal Error. Aborting at EstimatorManager input:unparsable <estimator> node, name: density type: density in Estimators input.`
Has the `density` estimator been ported to the batched code? If not, should I keep it or comment it out for now?

**Comment by prckent** [comment]
I think we decided that since the spindensity estimator was a superset of the density estimator then we didn't need both. However, the error message suggests that is not what is going on here (?).  At the very least the error message needs improving. There are other estimators that are close to duplicates of functionality, e.g. the various structure factors.

**Comment by aannabe** [comment]
Looking at `src/Estimators/EstimatorManagerInput.cpp` there is no if-clause for the density estimator. It looks like it hasn't been ported. I have removed that estimator from the example to avoid confusion. I have also always used the spin-density estimator to get the total charges.

**Comment by jtkrogel** [comment]
This is good.  Thanks Gani!

If you have time, can you also add the momentum distribution?  This is the other one that should have support in the batched drivers already.

**Comment by aannabe** [comment]
Will add the momentum distribution sometime today.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        warmupsteps      = 20,
+        blocks           = 20,
+        steps            = 20,
+        nonlocalmoves    = 'v3',
+        ),
+    dmc( # main dmc
+        walkers_per_rank = 8,
+        timestep         = 0.005,
+        warmupsteps      = 100,
+        blocks           = 100,
```
Comment (jtkrogel) [inline nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py:211]: Favor more blocks and fewer steps in all of these.  We are turning observables on in this example and we should demonstrate here that people should use many blocks for robust statistical analysis afterward.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    input_dft        = 'lda',
+    ecutwfc          = 400,
+    bandfac          = 1.5,    # Adjust as needed for the 1RDM orbital basis
+    occupations      = 'smearing',
+    smearing         = 'fermi-dirac',
+    degauss          = 1e-3,
+    mixing_beta      = 0.4,
+    nspin            = 2,
+    U_projection_type = 'ortho-atomic',
+    hubbard_u        = obj(Fe=2.0),
```
Comment (jtkrogel) [inline nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py:38]: Should be updated for compatibility with newer QE versions.  See Kayahan's PR.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    spin_polarized = True,
+    pseudos        = pseudos,
+    )
+
+# read the primitive cell
+structure = read_structure('Fe_bcc.xsf')
+structure.change_units('B')
+
+system = generate_physical_system(
+    structure  = structure.copy(),
```
Comment (jtkrogel) [inline nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py:58]: Please replace this line with `structure = 'Fe_bcc.xsf'` and remove the structure lines above.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+)
+
+qmc = generate_qmcpack(
+    #skip_submit          = True,
+    identifier           = 'qmc',
+    path                 = basepath + 'qmc',
+    job                  = job(cores=12,threads=4,app='qmcpack'),
+    system               = system,
+    twistnum             = 0,
+    estimators           = [sdens, edens, momdist, dm_est],  # Requested estimators. These will be included in all QMC series.
```
Comment (jtkrogel) [inline nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py:190]: Do these not appear in an `<estimators>` XML block in the input file (batched drivers)?  If not, then I need to make an update.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    twistnum             = 0,
+    qmc                  = 'opt',
+    J2                   = True,
+    minmethod            = 'oneshift',
+    warmupsteps          = 10,
+    init_cycles          = 5,
+    cycles               = 5,
+    walkers_per_rank     = 256,
+    samples              = 1e5,
+    blocks               = 10,
```
Comment (jtkrogel) [inline nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py:113]: Favor more blocks here so people can make a statistical analysis of the optimization afterwards.

**Review by jtkrogel — CHANGES_REQUESTED** [review]
See comments

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py (hunk 1 above)
Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py]: Increased the number of blocks. I kept the number of steps the same to follow the rule of thumb of "1/timestep steps for each block". Also due to @prckent 's comment below about I/O efficiency.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py (hunk 3 above)
Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py]: Done

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py (hunk 4 above)
Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py]: They do properly appear in a dedicated `<estimators>` XML block. My initial comment here tried to convey that all estimators will be run in all QMC series. Slightly updated the comment.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py (hunk 5 above)
Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py]: Done

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py (hunk 1 above)
Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py]: Also added a comment right above the production QMC block about the number of blocks, steps, and disk space considerations.

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py (hunk 2 above)
Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py:38]: Somehow, I can't get it working with the new QE input style for DFT+U. Left @kayahans a message.

Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py]: This is also fixed now.

**Comment by aannabe** [comment]
It looks good to me now. Feel free to push to my branch directly if you think the included comments need refinining.

**Review by jtkrogel — APPROVED** [review]
LGTM

File: nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+# In production runs containing estimators, large numbers (100+) of blocks are preferred for robust statistical analysis.
+# More steps per block than specified here are additionally preferred for computational efficiency due to the I/O cost of estimators.
+# Finally, note that for large numbers of blocks, the stat.h5 file's disk footprint can get considerably large.
+# This is especially true for dense grids, say spindensity with (300x300x300).
+qmc = generate_qmcpack(
+    #skip_submit          = True,
+    identifier           = 'qmc',
+    path                 = basepath + 'qmc',
+    job                  = job(cores=12,threads=4,app='qmcpack'),
```
Comment (prckent) [inline nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py:188]: Noticed you are using 12 cores here but 16 elsewhere. Any specific reason?

Comment (aannabe) [reply nexus/examples/qmcpack/rsqmc_misc/estimators/iron_ldaU_dmc.py]: Now changed to 16 as well. I must've copied that section from another example.

**Review by prckent — COMMENTED** [review]
I ran this workflow after changing generate_only to 0. The QMC run with the estimators failed. I will poke more later, but did it work for you? If so, which QMCPACK git hash/version?

```
Fatal Error. Aborting at EstimatorManager input:unparsable <estimator> node, name: edcell type: energydensity in Estimators input.
```

relevant input section

```
    <estimators>
         <estimator type="spindensity" name="SpinDensity">
            <parameter name="dr">
               0.05 0.05 0.05
            </parameter>
         </estimator>
         <estimator type="EnergyDensity" name="EDcell" dynamic="e" static="ion0">
            <spacegrid coord="cartesian">
               <origin p1="zero"/>
               <axis p1="a1" scale=".5" label="x" grid="-1 (100) 1"/>
               <axis p1="a2" scale=".5" label="y" grid="-1 (100) 1"/>
               <axis p1="a3" scale=".5" label="z" grid="-1 (100) 1"/>
            </spacegrid>
         </estimator>
         <estimator type="MomentumDistribution" name="nofk" samples="40" kmax="8.0"/>
         <estimator type="OneBodyDensityMatrices" name="DensityMatrices">
            <parameter name="energy_matrix"       >    no                    </parameter>
            <parameter name="integrator"          >    density               </parameter>
            <parameter name="basis"               >
               spo_dm
            </parameter>
            <parameter name="evaluator"           >    matrix                </parameter>
            <parameter name="check_overlap"       >    no                    </parameter>
            <parameter name="samples"             >    10                    </parameter>
         </estimator>
      </estimators>
```

**Comment by prckent** [comment]
Working now after an update & rebuild.

**Comment by aannabe** [comment]
Thanks. I also did a short VMC run on this system and it runs with the latest develop. The stat.h5 has all the included estimators:
```
ElecElec                 Group
IonIon                   Group
Kinetic                  Group
LocalECP                 Group
LocalEnergy              Group
LocalEnergy_sq           Group
LocalPotential           Group
MPC                      Group
NEEnergyDensityEstimator Group
NonLocalECP              Group
OneBodyDensityMatrices   Group
SpinDensity              Group
nofk                     Group
```

PS. The new DFT+U input style using QE v7.4 also works.

**Review by prckent — APPROVED** [review]
Thanks Gani.

**Comment by prckent** [comment]
Test this please

**Pull request merged**